Compiler/script2/cs_parser.cpp

```diff
@@ -2557,6 +2562,89 @@ void AGS::Parser::ParseExpression_Ternary(size_t const tern_idx, SrcList &expres
+
+    Vartype const curr_vartype = eres.Vartype;
+
+    if (_sym.IsDynpointerVartype(curr_vartype))
+    {
+        if (!_sym.IsDynpointerVartype(new_vartype))
+            new_vartype = _sym.VartypeWithDynpointer(new_vartype);
+    }
+
+    // Handle cases where 'curr_vartype' is convertable to 'new_vartype'
+    // even without any 'as' clause.
+    if (!IsVartypeMismatch_Oneway(curr_vartype, new_vartype))
+    {
+        eres.Vartype = new_vartype;
+        return;
+    }
+
+    if (_sym.IsDynarrayVartype(curr_vartype))
+        // Here, when we try to convert a (managed) dynarray type
+        // 'A*[]' to type 'B*[]' where 'B' is not an ancester of 'A':
+        // If we'd do this, then the compiler would "forget"
+        // that the array used to have the type 'A*[]' originally.
+        // At this point in time, this array might already contain
+        // non-null elements. We will never run checks whether these
+        // actually _are_ of type 'B*'.
+        // This is unsafe. So do NOT allow this.
+        UserError(    
+            ReferenceMsgSym(
+                "Cannot convert the dynamic array '%s' to type '%s'",
+                curr_vartype).c_str(),
+            _sym.GetName(curr_vartype).c_str(),
+            _sym.GetName(new_vartype).c_str());
+
+    if (_sym.IsAtomicVartype(curr_vartype))
+        // Here, we might implement an int-to-float or float-to-int
+        // conversion in future. There are several possibilities
+        // for the float-to-int conversion (to nearest, lower,
+        // upper int, towards zero, away from zero), so which one?
+        UserError(
+            "Cannot convert type '%s' to type '%s' through an 'as' clause",
+            _sym.GetName(curr_vartype).c_str(),
+            _sym.GetName(new_vartype).c_str());
+
+    if (!_sym.IsDynpointerVartype(curr_vartype))
+        InternalError(
+            "Expected to convert a dynpointer, found '%s'",
+            _sym.GetName(curr_vartype).c_str());
+
+    Vartype const element_vartype = _sym.GetFirstBaseVartype(new_vartype);
+    if (!_sym.IsManagedVartype(element_vartype))
+        UserError(
+            ReferenceMsgSym(
+                "Cannot convert to the non-managed type '%s'",
+                new_vartype).c_str(),
+            _sym.GetName(new_vartype).c_str());
+
+    if (FlagIsSet(_options, SCOPT_RTTIOPS))
+    {
+        EvaluationResultToAx(eres);
+        PushReg(SREG_MAR);                                      // TODO: remove when SCMD_DYNAMICCAST processes AX
+        WriteCmd(SCMD_REGTOREG, SREG_AX, SREG_MAR);   // TODO: remove when SCMD_DYNAMICCAST processes AX
+        CodeCell const scmd_dynamiccast = 76;               // TODO: remove when SCMD_DYNAMICCAST is defined
+        WriteCmd(scmd_dynamiccast, element_vartype); // TODO: change to SCMD_DYNAMICCAST when that is defined
+        WriteCmd(SCMD_REGTOREG, SREG_AX, SREG_MAR);   // TODO: remove when SCMD_DYNAMICCAST processes AX
```

Comment:
I suspect there's a mistake here, it should be reverse
```WriteCmd(SCMD_REGTOREG, SREG_MAR, SREG_AX);```
otherwise the SREG_AX never changes, and it looks like any cast succeeds unconditionally.